### PR TITLE
Replace Lambda function location_from_latlon with stored generated column

### PR DESCRIPTION
**Pull request by roseeichelmann**
## Associated issues
Closes https://github.com/cityofaustin/atd-data-tech/issues/13055

~This PR recreates the `location_id` column but as a stored generated column for easier maintenance in keeping the column up to date whereas currently we rely on SQS and AWS lambda functions to do so. Related to Mikes work in https://github.com/cityofaustin/atd-vz-data/pull/1269 and I think we will have to coordinate deployment together to get this done most efficiently, bc like Mike talks about there are lots of views that depend on the `location_id` of cr3 crashes and noncr3 crashes.~

This PR now uses a trigger function to keep the location_id column up to date for the atd_txdot_crashes table.

## Testing
**URL to test:** 
Local db

**Steps to test:**

Using some of Mikes test/deployment steps here so shoutout Mike for coming up with great ways to test this!

1. Follow the deployment steps below in your local db instance
2. Test that the insert trigger works by running the following commands:

```plpgsql
-- insert a cr3 crash with coordinates that will associate to location_id 8337406736
INSERT INTO atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (30.16841600, -97.78449000, 9999);

-- insert a cr3 crash with coordinates that will not associate to a location
INSERT INTO atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (30.386136, -97.67309, 8888);

-- insert a cr3 crash without coordinates that will not associate to a location
INSERT INTO atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (NULL, NULL, 7777);
``` 
3. Update an existing record with a location_id of NULL to have the `latitude_primary` and `longitude_primary` (these are the column values that are updated when a user uses the crash UI map to update coordinates) of a record with a location. After update, you should see the record associate with the same location as the record you took the lat/lon from.
4. Update an existing record with a location_id that is not null and update it with `latitude_primary` and `longitude_primary` coordinates from an existing record that has latitude and longitude and no location_id associated.

## Deployment steps

- [ ] Remove the Hasura trigger called location_from_latlon by going on to the Hasura console and go to the Events tab.
- [ ] Run SQL command to create the trigger function in [update_cr3_location.sql](https://github.com/cityofaustin/atd-vz-data/blob/dd674a85aca828307bf751784575822d06dbb263/atd-vzd/triggers/update_cr3_location.sql)
- [ ] Run SQL commands in the migrations file [migration_atd_txdot_crashes_2023_08_03--1127.sql](https://github.com/cityofaustin/atd-vz-data/blob/dd674a85aca828307bf751784575822d06dbb263/atd-vzd/migrations/migration_atd_txdot_crashes_2023_08_03--1127.sql)
- [ ] Switch off cr3 crash location housekeeping Airflow ETLs

---
#### Ship list
- [ ] Code reviewed
- [ ] Product manager approved

### CODE REVIEW

File: atd-events/crash_update_jurisdiction/README.md
Comment (roseeichelmann) [inline atd-events/crash_update_jurisdiction/README.md:1]: look at all that red :heart_eyes:

File: atd-vzd/triggers/is_crash_level_5.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+-- This function determines whether crash occurred on a level 5 road (large highway)
+-- and is thus outside of COA jurisdiction.
+CREATE OR REPLACE FUNCTION is_crash_level_5(crash_id integer, rpt_road_part_id integer, rpt_hwy_num text) RETURNS boolean AS $$
+        BEGIN
+        RETURN (
+            SELECT(rpt_road_part_id != 2 AND UPPER(LTRIM(rpt_hwy_num)) IN ('35', '183','183A','1','290','71','360','620','45','130'))
```
Comment (roseeichelmann) [inline atd-vzd/triggers/is_crash_level_5.sql:6]: based on logic from https://github.com/cityofaustin/atd-data-tech/issues/11368 where xavier defines what a level 5 crash is

File: atd-vzd/triggers/get_cr3_location_id.sql [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+                RETURN NULL;
+            -- Return the location id of the crash by finding which location polygon the crash
+            -- geographic position resides in
+            ELSE
+                RETURN (
+                    SELECT location_id 
+                    FROM atd_txdot_locations 
+                    WHERE location_group = 1 
+                    AND (geometry && location)
+                    AND ST_Contains(geometry, location)
```
Comment (roseeichelmann) [inline atd-vzd/triggers/get_cr3_location_id.sql:17]: todo is to make sure `geometry` is what we want to use instead of `shape` and then drop the shape column from `atd_txdot_locations`

File: atd-vzd/migrations/migration_atd_txdot_crashes_2023_08_03--1127.sql [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+-- TODO drop DB views that use location_id (replace later with generated location_id)
+
+-- Drop current atd_txdot_crashes location_id column
+ALTER TABLE atd_txdot_crashes DROP COLUMN location_id;
+
+-- Rename generated_location_id to location_id to replace it
+ALTER TABLE atd_txdot_crashes 
+RENAME COLUMN generated_location_id TO location_id;
+
+-- TODO add DB views that use location_id
```
Comment (roseeichelmann) [inline atd-vzd/migrations/migration_atd_txdot_crashes_2023_08_03--1127.sql:15]: following pattern from https://github.com/cityofaustin/atd-vz-data/pull/1269 for steps to retire the old column and rename the new generated column

File: atd-vzd/triggers/get_cr3_location_id.sql [hunk 4]
```diff
@@ -0,0 +1,21 @@
+-- This function is invoked by a stored generated column called location_id in the atd_txdot_crashes table.
+-- It is responsible for finding the location id of a crash, setting the location_id to NULL if the crash occurred on a 
+-- level 5 roadway and thus is outside of COA jurisdiction.
+CREATE OR REPLACE FUNCTION get_cr3_location_id(location geometry(Geometry,4326), crash_id integer, rpt_road_part_id integer, rpt_hwy_num text) RETURNS text AS $$
+        BEGIN
+            -- If crash is level 5 return null because it is outside of COA jurisdiction
+            IF (is_crash_level_5(crash_id, rpt_road_part_id, rpt_hwy_num)) THEN
```
Comment (roseeichelmann) [inline atd-vzd/triggers/get_cr3_location_id.sql:7]: instead of putting the logic for determining if a crash is level 5 within the code here, i decided to make a diff function that does that in case this logic needs to be reused again elsewhere in the future? lmk if yall think thats right. also lmk if you think there is a better descriptive name than `is_crash_level_5`

File: atd-vzd/triggers/is_crash_level_5.sql (hunk 1 above)
Comment (roseeichelmann) [reply atd-vzd/triggers/is_crash_level_5.sql:6]: something i just thought of tho - what if the rpt_road_part_id or rpt_hwy_num is wrong and since those arent values the VZ team can fix/update, then the crash is just wrongfully identified as level 5 and isnt given a location?

**Comment by frankhereford** [comment]
No worries, but I think you may have assigned us when you were ready for reviewers.

**Comment by roseeichelmann** [comment]
@frankhereford whoops, i did indeed!

File: atd-vzd/migrations/migration_atd_txdot_crashes_2023_08_03--1127.sql (hunk 3 above)
Comment (mddilley) [reply atd-vzd/migrations/migration_atd_txdot_crashes_2023_08_03--1127.sql:15]: cool! fyi, I'm going to fill in these TODOs in my PR on Monday and then request review. I felt way to scattered today to untangle the dependencies that block dropping the current non-CR3 location column.

Comment (roseeichelmann) [reply atd-vzd/migrations/migration_atd_txdot_crashes_2023_08_03--1127.sql:15]: @mddilley okay sounds good, i figured id just post the rest of my code for review now but ill also look into what it will take to drop the old column :-)

File: atd-vzd/triggers/is_crash_level_5.sql (hunk 1 above)
Comment (mddilley) [reply atd-vzd/triggers/is_crash_level_5.sql:6]: I think that it is. When we talked with Xavier in the past, he imagined being able to edit those columns, but, when Frank and I synced on it afterward, we thought about another editable boolean column like `is_level_5` that they could edit to push a crash into the accurate category. Hoping that we can cover that in the meeting on Monday morning to get a feel for it!

**Comment by roseeichelmann** [comment]
I updated the function for the generated column to be:

image

like we talked about in our meeting today, but when i delete the generated column and create it again i get the error "more than one row returned by a subquery used as an expression" and i think its because that first return where `location_group = 2` is maybe returning multiple records in some cases. Is it possible that there are overlapping `location_group = 2` polygons?

Gonna do more investigating on this

**Comment by johnclary** [comment]
> Is it possible that there are overlapping location_group = 2 polygons

Overlapping polygons is obviously not ideal, but for the purpose of making sure the column doesn't break, maybe you should just add a `LIMIT 1` to each of the `SELECT` statements?

File: atd-vzd/triggers/get_cr3_location_id.sql (hunk 4 above)
Comment (roseeichelmann) [reply atd-vzd/triggers/get_cr3_location_id.sql:10]: im kinda lost on what terminology we are going with. "level 5", "mainlane", and "major road" are thrown around kinda interchangeably in our convos and it would be good to maybe pick one for describing these kinds of roads. i think we already discussed how mainlane isnt super accurate bc small neighborhood streets also have mainlanes, so either "level 5" or "major road" sounds good to me.

**Comment by roseeichelmann** [comment]
Views that use the `location_id` column of the `atd_txdot_crashes` table:

atd_txdot_crash_locations_ranking
cr3_nonproper_crashes_on_mainlane
locations_with_crash_injury_counts
view_atd_crashes_full_purpose
view_location_crashes_global
view_location_injry_count_cost_summary

materialized views:
all_atd_txdot_crashes
all_cr3_crashes_off_mainlane
five_year_atd_txdot_crashes
five_year_cr3_crashes_off_mainlane
five_year_cr3_crashes_outside_surface_polygons

I think we should have a convo about which, if any, of these views we can drop if they aren't used anymore like mike mentioned in https://github.com/cityofaustin/atd-vz-data/pull/1269. some of them are views of other views - viewception :hot_face:

File: atd-vzd/migrations/migration_atd_txdot_crashes_2023_08_03--1127.sql [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+FOR EACH ROW
+WHEN (OLD.position IS DISTINCT FROM NEW.position)
+EXECUTE PROCEDURE update_cr3_location();
+
+COMMENT ON FUNCTION update_cr3_location IS 'This trigger function is used to update the location_id on insert and update of a record in atd_txdot_crashes.';
+
+-- Drop functions that are no longer needed
+DROP FUNCTION find_cr3_mainlane_crash;
+DROP FUNCTION find_location_for_cr3_collision;
+DROP FUNCTION find_location_id_for_cr3_collision;
```
Comment (frankhereford) [inline atd-vzd/migrations/migration_atd_txdot_crashes_2023_08_03--1127.sql:19]: If you want, you can stick a linefeed on here, but this file will probably get zapped in an upcoming PR from Mike with VZ migration support, so all good.

File: atd-vzd/migrations/migration_atd_txdot_crashes_2023_06_26--1622.sql [hunk 6]
```diff
@@ -1,2 +1,2 @@
 ALTER TABLE atd_txdot_crashes
-ADD COLUMN in_austin_full_purpose boolean GENERATED ALWAYS AS (is_crash_in_austin_full_purpose(position, city_id)) STORED NOT NULL;
```
Comment (roseeichelmann) [inline atd-vzd/migrations/migration_atd_txdot_crashes_2023_06_26--1622.sql:2]: realized this needs to be nullable, was having some error issues when testing adding a new crash record that had null for lat and long. already made the column nullable in staging/prod so this just reflects those changes

**Review by frankhereford — APPROVED** [review]
This works great, and you have some really helpful instructions and background in your PR. This was an awesome review experience. 

One thing I ran into was that when I followed your testing instructions and looked at the location ID that was assigned to the crash `9999`, I saw that it was all integers, so I worried that you may be pulling over the integer version of the polygon ID instead of the hex one, but .. I did this testing:

```
-- insert a cr3 crash with coordinates that will associate to location_id 8337406736
INSERT INTO atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (30.16841600, -97.78449000, 9999);

-- insert a cr3 crash with coordinates that will not associate to a location
INSERT INTO atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (30.386136, -97.67309, 8888);

-- insert a cr3 crash without coordinates that will not associate to a location
INSERT INTO atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (NULL, NULL, 7777);


select crash_id, location_id from atd_txdot_crashes where crash_id in (9999,8888,7777);

select crash_id, location_id from atd_txdot_crashes where location_id is not null limit 10;

select location_id, polygon_hex_id from atd_txdot_locations where location_id = '8337406736';
```
... and it turns out that this just happens to land in polygon identified by a hex number where all of its digits are 0-9, which is really kind of remarkable.

Anyway - this is 🚢 🚢 🚢 🚢 !

File: atd-vzd/migrations/migration_atd_txdot_crashes_2023_06_26--1622.sql (hunk 6 above)
Comment (mddilley) [reply atd-vzd/migrations/migration_atd_txdot_crashes_2023_06_26--1622.sql:2]: awesome - just noting that your change was reflected in the giant migration™️ 🐘

File: atd-vzd/triggers/update_cr3_location.sql [hunk 7]
```diff
@@ -0,0 +1,29 @@
+-- This trigger function updates the location_id of a record in atd_txdot_crashes on insert or update.
+CREATE OR REPLACE FUNCTION update_cr3_location()
+    RETURNS TRIGGER AS $$
+        BEGIN
+            -- If crash is level 5 try to associate crash to level 5 polygon and return the location id
```
Comment (mddilley) [inline atd-vzd/triggers/update_cr3_location.sql:5]: Sorry to dig this back up, but are we sure that the VZ team is on board with using the current level 5 polygons? Or, we might be holding off on this until the new level 5s are ready for production.

**Review by mddilley — APPROVED** [review]
I followed through the test steps, and I'm seeing location's update as expected! 👏 I'm going to take some tomorrow to 1. get my branch up to date from master and 2. branch from `md-13057-blueform-lat-lon-trigger` to create a separate PR to create Hasura migrations for my changes.

Thanks for working through this and teaching us along the way. We will have this worked all merged in no time! 🙌

File: atd-vzd/triggers/update_cr3_location.sql (hunk 7 above)
Comment (roseeichelmann) [reply atd-vzd/triggers/update_cr3_location.sql:5]: This is what Xavier wanted me to change this code to in our last meeting soo i think so :shrug:

**Review by chiaberry — APPROVED** [review]
I saw that there are conflicts with one of the migration files, because the master branch has changed since this one was started -- sorry I was so slow to review this PR. Does the migration file need to be renamed?

I do love all the red.

**Review by patrickm02L — APPROVED** [review]
Adding my ✅  based on the Dev reviews!
